ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,668 @@
+`2` as described by the data inside `L11`. Finally, processing starts again at
+`L12`, during which the Process BPMN element of workflow `1` in workflow
+instance `2` transitions to `Element_Activated`.
+
+Now, let's again make the same imaginary bug fix in the key generation. From
+now, only odd numbers are generated.
+
+Updating to this new version from `S1` with replay starting at `L10` results is
+the same state as before the update. First `L10` is skipped, because commands
+are skipped during replay. Then, during the replay of `L11`, a workflow instance
+is stored in the state with key `2` as described by the data inside `L11`. Note,
+that this key is not newly generated, in which case it would've had the value
+`3`. Finally, processing starts again at `L12`, during which the Process BPMN
+element of workflow `1` in workflow instance `2` transitions to
+`Element_Activated`.
+
+Note, that in this case the workflow instance was correctly referred to from
+`L12`, because the events were directly applied as changes to the state instead of reapplying the business logic (which had changed).
+
+This example also shows that the effect of this proposal is that bug fixes in
+the form of business logic changes can no longer be applied to the processing of
+records that have already been processed previously. Their situation will not be
+altered since their records have already been written to the log stream in the
+form of events.
+ 
+## Key Generation
+
+The failure in the previous example occurred in key generation. Even though this
+proposal improves dealing with changes to keys, we should consider keys in more
+detail.
+
+A key (i.e. a long value) identifies an entity (e.g. a workflow instance) uniquely across all partitions. Most records have a key, except for user commands that don't refer to an existing resource (e.g. create workflow instance command). When a record is written for a new entity when processing a command, then a new key is generated.
+
+On replay, the key generator must be restored to generate new keys on processing. To restore the key generator, we read all records on the stream to find the latest key (i.e. the highest long value). Note, that the last record on the stream might not hold the latest key, because it can refer a previous entity.
+
+The restoring requires that all generated keys are also stored in records on the stream (instead of only using the key as data in the state).
+
+## State Transitions as Commands
+
+The transition from one state of an element instance to the next state (i.e. `Element_Activating`, `Element_Completing`, `Element_Terminating`) is modeled as a command. 
+
+![state-transistions](ZEP-wf-stream-processing/event-sourcing-state-transitions.png)
+
+A state transition can be interrupted by a concurrent command (e.g. canceling the workflow instance). In this case, the processor rejects the command and doesn't apply its business logic. Command rejections are discussed in more detail later.
```

Comment:
The log stream is processed in sequence, but there are multiple writers appending new records (including commands) to the log stream. Concurrent command relates to a command written concurrently to the processing of the stream. Such commands can disrupt the processing flow leading to command rejections. The example mentions cancelling the workflow instance. Does that make sense or should we explain this in more detail?

Reply:
It does make sense and I think it would help to explain it in more detail. Especially because an event stream log is such a serial data struture :-) 

Doesn't have to be long. Might be as simple as defining in the beginning
"By concurrent commands we mean commands that are being written (by other log stream writers) while events are being processed. Such commands can disrupt the processing flow leading to command rejections." Maybe it would help to clarifiy whether the disruption/rejection affects commands that are currently being processed, or that have been written to log but not processed yet (which I think is what happen), or whether it only effects stuff that has not yet been written to the log.

Reply:
Yes. Let's describe it a bit more in the section :+1: